### PR TITLE
A0-2711: update ink version

### PR DESCRIPTION
**Pull request by fbielejec**
# Description

ink! 4.1.2 version update

### CODE REVIEW

File: .github/workflows/build-and-push-cliain.yml [hunk 1]
```diff
@@ -6,6 +6,8 @@ on:
     branches:
       - 'main'
   pull_request:
+    paths:
+      - bin/cliain/**
```
Comment (pmikolajczyk41) [inline .github/workflows/build-and-push-cliain.yml:10]: yes, but ideally we would have also a check that `Cargo.lock`s are up-to-date - it used to be pretty common to make changes only to `aleph-client`, which actually should affect building `cliain`, which requires bumping `aleph-client` and updating `Cargo.lock` in `cliain`

not a blocker, just raising warning

File: contracts/adder/Cargo.toml [hunk 2]
```diff
@@ -5,19 +5,14 @@ authors = ["[your_name] <[your_email]>"]
 edition = "2021"
 
 [dependencies]
-anyhow = "*"
-ink = { version = "4.2.0",  default-features = false }
+ink = { version = "4.2",  default-features = false }
```
Comment (pmikolajczyk41) [inline contracts/adder/Cargo.toml:8]: I'd keep `ink` versions identical in all contracts: e.g. in `access_control` we have `=4.2.1`, which seems legit to me

File: contracts/scripts/deploy.sh [hunk 3]
```diff
@@ -1,9 +1,11 @@
 #!/bin/bash
 
-set -euox pipefail
+# set -x
```
Comment (pmikolajczyk41) [inline contracts/scripts/deploy.sh:3]: leftover?

File: contracts/adder/Cargo.toml (hunk 2 above)
Comment (fbielejec) [reply contracts/adder/Cargo.toml:8]: done

File: .github/workflows/build-and-push-cliain.yml (hunk 1 above)
Comment (fbielejec) [reply .github/workflows/build-and-push-cliain.yml:10]: I understand, but the bulk of the time on this PR was already spent on task not related to the original issue. 
Just goes on to show there's still some work left to do on the pipelines

File: contracts/scripts/deploy.sh (hunk 3 above)
Comment (fbielejec) [reply contracts/scripts/deploy.sh:3]: I understand -x as a debugging flag, as it is rather noisy. I can remove or leave it commented out

File: .github/workflows/_check-excluded-packages.yml [hunk 4]
```diff
@@ -43,49 +46,52 @@ jobs:
...
+          rustup toolchain install ${RUST_CONTRACTS_TOOLCHAIN}
+          rustup toolchain install ${RUST_TOOLCHAIN}
+          cargo +${RUST_CONTRACTS_TOOLCHAIN} install cargo-dylint dylint-link
+          cargo +${RUST_CONTRACTS_TOOLCHAIN} install --version $CARGO_CONTRACT_VERSION \
+            --force --locked cargo-contract
 
-          rustup component add rust-src --toolchain nightly-2022-10-30-x86_64-unknown-linux-gnu
-          cargo install cargo-dylint dylint-link
-          cargo install --version 2.1.0 --force --locked cargo-contract
+          # packages="${{ steps.format_output.outputs.packages }}"
```
[the remark this answers is not in the document]
Comment (fbielejec) [reply .github/workflows/_check-excluded-packages.yml]: this entire workflow is re-written, I think you saw it mid-changes

**Review by pmikolajczyk41 — APPROVED** [review]

File: scripts/run_checks_on_excluded_packages.sh [hunk 5]
```diff
@@ -2,29 +2,50 @@
...
-  "contracts/simple_dex"
-  "contracts/ticket_token"
-  "contracts/wrapped_azero"
-  "contracts/adder, "
-)
-
-for p in ${packages[@]}
-do
+TOML_FILE="Cargo.toml"
+RUST_TOOLCHAIN=nightly-2022-10-30-x86_64-unknown-linux-gnu
```
Comment (mike1729) [inline scripts/run_checks_on_excluded_packages.sh:6]: Maybe we should read the value from `rust-toolchain` file?

File: scripts/run_checks_on_excluded_packages.sh [hunk 6]
```diff
@@ -2,29 +2,50 @@
...
+
+# Remove the key
+packages=${packages:10}
+
+for p in ${packages[@]}; do
+
   echo "Checking package $p..."
   pushd "$p"
+
+  if [ $p = "contracts/adder" ] ||
```
Comment (mike1729) [inline scripts/run_checks_on_excluded_packages.sh:27]: You can simplify this if condition to `if [[ $p = "contracts/"* ]] && [ p != "contracts/poseidon_host_bench" ] ; then`

File: scripts/run_checks_on_excluded_packages.sh [hunk 7]
```diff
@@ -2,29 +2,50 @@
...
+       [ $p = "contracts/button" ] ||
+       [ $p = "contracts/game_token" ] ||
+       [ $p = "contracts/marketplace" ] ||
+       [ $p = "contracts/simple_dex" ] ||
+       [ $p = "contracts/ticket_token" ] ||
+       [ $p = "contracts/wrapped_azero" ] ||
+       [ $p = "contracts/shared_traits" ] ; then
+    cargo +${RUST_CONTRACTS_TOOLCHAIN} contract check
+  elif [ $p = "baby-liminal-extension" ] || [ $p = "contracts/poseidon_host_bench" ]; then
+    # cargo clippy --release --no-default-features --features substrate \
```
Comment (mike1729) [inline scripts/run_checks_on_excluded_packages.sh:39]: why this is commented?

**Review by mike1729 — COMMENTED** [review]
Just a couple comments.

File: scripts/run_checks_on_excluded_packages.sh (hunk 7 above)
Comment (fbielejec) [reply scripts/run_checks_on_excluded_packages.sh:39]: no idea, inherited like this maybe @pmikolajczyk41 knows

Comment (pmikolajczyk41) [reply scripts/run_checks_on_excluded_packages.sh:39]: poseidon-host-bench should be removed iirc, baby-liminal-extension is not migrated yet (I'm waiting for either dev need for it or crates.io publishing of obce)

File: scripts/run_checks_on_excluded_packages.sh (hunk 6 above)
Comment (fbielejec) [reply scripts/run_checks_on_excluded_packages.sh:27]: great idea, done

File: scripts/run_checks_on_excluded_packages.sh (hunk 5 above)
Comment (fbielejec) [reply scripts/run_checks_on_excluded_packages.sh]: good idea, added

**Review by mike1729 — APPROVED** [review]

**Pull request merged**